hare3/hare.go

```diff
@@ -0,0 +1,571 @@
+			)
+			input.result <- &response{gossip: gossip, equivocation: equivocation}
+		case <-h.wallclock.After(h.wallclock.Until(walltime)):
+			h.log.Debug("execute round",
+				zap.Uint32("lid", layer.Uint32()),
+				zap.Uint8("iter", proto.Iter), zap.Stringer("round", proto.Round),
+			)
+			current := proto.IterRound
+			start := time.Now()
+			var vrf *types.HareEligibility
+			if current.IsMessageRound() {
+				vrf = h.oracle.active(h.signer.NodeID(), layer, current)
+				activeLatency.Observe(time.Since(start).Seconds())
+			}
+			h.tracer.OnActive(vrf)
+
+			out := proto.next(vrf != nil)
+			if err := h.onOutput(layer, current, out, vrf); err != nil {
+				return err
+			}
+			if out.terminated {
+				return nil
+			}
+			if current.Iter == h.config.IterationsLimit {
+				return fmt.Errorf("hare failed to reach consensus in %d iterations",
+					h.config.IterationsLimit)
+			}
+			walltime = walltime.Add(h.config.RoundDuration)
+		case <-h.ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (h *Hare) onOutput(layer types.LayerID, ir IterRound, out output, vrf *types.HareEligibility) error {
+	h.log.Debug("round output",
+		zap.Uint32("lid", layer.Uint32()),
+		zap.Uint8("iter", ir.Iter), zap.Stringer("round", ir.Round),
+		zap.Inline(&out),
+		zap.Bool("active", vrf != nil),
+	)
+	if out.message != nil {
+		h.eg.Go(func() error {
+			out.message.Layer = layer
+			out.message.Eligibility = *vrf
+			out.message.Sender = h.signer.NodeID()
+			out.message.Signature = h.signer.Sign(signing.HARE, out.message.ToMetadata().ToBytes())
+			if err := h.pubsub.Publish(h.ctx, h.config.ProtocolName, out.message.ToBytes()); err != nil {
+				h.log.Error("failed to publish", zap.Inline(out.message))
+			}
+			h.tracer.OnMessageSent(out.message)
+			return nil
+		})
+	}
+	if out.coin != nil {
+		select {
+		case <-h.ctx.Done():
+			return h.ctx.Err()
+		case h.coins <- WeakCoinOutput{Layer: layer, Coin: *out.coin}:
+		}
+		sessionCoin.Inc()
+	}
+	if out.result != nil {
+		select {
+		case <-h.ctx.Done():
+			return h.ctx.Err()
+		case h.results <- ConsensusOutput{Layer: layer, Proposals: out.result}:
+		}
+		sessionResult.Inc()
+	}
+	return nil
+}
+
+func (h *Hare) Proposals(lid types.LayerID, epochBeacon types.Beacon) []types.ProposalID {
+	h.log.Debug("requested proposals",
+		zap.Uint32("lid", lid.Uint32()),
+		zap.Stringer("beacon", epochBeacon),
+	)
+	props, err := proposals.GetByLayer(h.db, lid)
+	if err != nil {
+		if errors.Is(err, sql.ErrNotFound) {
+			h.log.Warn("no proposals found for hare, using empty set",
+				zap.Uint32("lid", lid.Uint32()), zap.Error(err))
+		} else {
+			h.log.Error("failed to get proposals for hare",
+				zap.Uint32("lid", lid.Uint32()), zap.Error(err))
+		}
+		return []types.ProposalID{}
+	}
+	var (
+		beacon types.Beacon
+		result []types.ProposalID
+	)
+	own, err := h.db.GetEpochAtx(lid.GetEpoch()-1, h.signer.NodeID())
+	if err != nil && !errors.Is(err, sql.ErrNotFound) {
+		h.log.Warn("no atxs in the requisted epoch",
```

Comment:
i actually copied this part from function in hare. i assumed that you wrote it this way so that it works for inactive smesher. but in my case this code won't be even executed for inactive, so i can change

Reply:
i changed code  to reflect that